ethcore/builtin/src/lib.rs

```diff
@@ -337,6 +368,62 @@ impl Implementation for Sha256 {
 	}
 }
 
+impl Implementation for Blake2F {
+	/// Format of `input`:
+	/// [4 bytes for rounds][64 bytes for h][128 bytes for m][8 bytes for t_0][8 bytes for t_1][1 byte for f]
+	fn execute(&self, input: &[u8], output: &mut BytesRef) -> Result<(), &'static str> {
+		const BLAKE2_F_ARG_LEN: usize = 213;
+		const PROOF: &str = "Checked the length of the input above; qed";
+
+		if input.len() != BLAKE2_F_ARG_LEN {
+			trace!(target: "builtin", "input length for Blake2 F precompile should be exactly 213 bytes, was {}", input.len());
+			return Err("input length for Blake2 F precompile should be exactly 213 bytes")
+		}
+
+		let mut cursor = Cursor::new(input);
+		let rounds = cursor.read_u32::<BigEndian>().expect(PROOF);
+
+		// state vector, h
+		let mut h = [0u64; 8];
+		for state_byte in h.iter_mut() {
+			*state_byte = cursor.read_u64::<LittleEndian>().expect(PROOF);
+		}
+
+		// message block vector, m
+		let mut m = [0u64; 16];
+		for msg_byte in m.iter_mut() {
+			*msg_byte = cursor.read_u64::<LittleEndian>().expect(PROOF);
+		}
+
+		// 2w-bit offset counter, t
+		let t = [
+			cursor.read_u64::<LittleEndian>().expect(PROOF),
+			cursor.read_u64::<LittleEndian>().expect(PROOF),
+		];
+
+		// final block indicator flag, "f"
+		let f = match input.last() {
+				Some(1) => true,
+				Some(0) => false,
+				_ => {
+					trace!(target: "builtin", "incorrect final block indicator flag, was: {:?}", input.last());
+					return Err("incorrect final block indicator flag")
+				}
+			};
+
+		compress(&mut h, m, t, f, rounds as usize);
+
+		output.write(0, unsafe {
+			// `from_raw_parts` is safe under the assumption that `h` is
+			// 1. immutable,
+			// 2. comprised of `u64`
+			// …and because we're using the (known) length of `h`
+			slice::from_raw_parts(h.as_ptr() as *const u8, h.len() * size_of::<u64>())
```

Comment:
It might be better to remove this unsafe. What I worry is someone compiling parity-ethereum to a big endian platform. There may not be many, but once people do that it breaks consensus. We may want to just specify the exact encoding how conversion from u64 array to bytearray happens.

Reply:
There could be other problems with big endian, e.g. with ethash https://github.com/paritytech/parity-ethereum/blob/5baa7e8fb50512b0de11b0994fa1b136de289831/ethash/src/compute.rs#L273-L276

but in this case it should work with big endian if I'm not mistaken